platform/platform-impl/src/com/intellij/openapi/editor/impl/EditorImpl.java

```diff
@@ -4085,19 +4085,42 @@ else if (e.getSource() != myGutterComponent && isCreateRectangularSelectionEvent
 
       if (moveCaret) {
         if (e.isShiftDown() && !e.isControlDown() && !e.isAltDown() && !e.isMetaDown()) {
+          int startToUse = 0;
           if (getMouseSelectionState() != MOUSE_SELECTION_STATE_NONE) {
             if (caretOffset < mySavedSelectionStart) {
-              mySelectionModel.setSelection(mySavedSelectionEnd, caretOffset);
+              startToUse = mySavedSelectionEnd;
             }
             else {
-              mySelectionModel.setSelection(mySavedSelectionStart, caretOffset);
+              startToUse = mySavedSelectionStart;
             }
           }
           else {
-            int startToUse = oldSelectionStart;
+            startToUse = oldSelectionStart;
             if (mySelectionModel.isUnknownDirection() && caretOffset > startToUse) {
               startToUse = Math.min(oldStart, oldEnd);
             }
+          }
+          if (isColumnMode()) {
+            final CaretModel caretModel = getCaretModel();
+            int selStart = oldStart;
+            int selEnd = oldEnd;
+            LogicalPosition blockStart;
+            if (caretModel.supportsMultipleCarets()) {
```

Comment:
We know for sure that our caret model supports multiple carets. This should be simplified.